boost/wave_1_71.py

```diff
@@ -0,0 +1,51 @@
+# coding: utf-8
+
+########################################
+# Wave 1.71
+########################################
+
+# Author: Jeff Trull
+
+from .utils import *
+
+@add_printer
+class BoostWaveFlexString:
+    """Pretty Printer for Boost.Wave internal string"""
+    printer_name = 'boost::wave::util::flex_string'
+    min_supported_version = (1, 71, 0)
+    max_supported_version = last_supported_boost_version
+    template_name = 'boost::wave::util::flex_string'
+
+    def __init__(self, val):
+        self.val = val
+
+    def to_string(self):
+        return call_object_method(self.val, 'c_str').string()
```

Comment:
Calling function from the binary should only be done as the last resort, if nothing else works. `flex_string::c_str()` can be absent from binary if it is not used in the program. Pretty printer will not work in that case.

I strongly advise to rewrite the printer in a way that uses underlying data members. Relying on standard library printers being available is ok.

Reply:
This one is a bit tricky, for several reasons:
1. The underlying data member is actually stored in a base class whose type is the third template parameter to `flex_string`. So in theory its type and name depends on that template parameter.
2. The implementation of `c_str()` in the default base class actually has a side effect of ensuring the data is null terminated. So simply returning the address of the first byte might not suffice. I might have to arrange for Python code to terminate it, or for some kind of copying, based on character count, into a local Python string.

While it's true that `c_str()` might not be called in the future, it is currently quite integral to the functioning of the parser, appearing in many places for generating error messages. Further, the use of data members might change in the future too, so that might not be any more stable.

If it helps, I am currently the most active maintainer of Wave, and I will probably remember this conversation if any plans to remove `c_str()` come up :)

Reply:
Oh, that's even worse! The implementation of c_str() can modify the state of the program. So, each time a value is printed, the memory of the program is silently changed. That's totally unacceptable. While I understand that data structure may change in future, it is not a way to go. Here is my implementation (of course, it is rather bulky and needs to be covered by a separate test).

```
def cast_array_to_pointer(val):
    """Cast array to pointer to the first element"""
    assert val.type.code == gdb.TYPE_CODE_ARRAY
    element_type = val.type.target()
    return val.address.cast(element_type.pointer())


@add_printer
class BoostWaveAllocatorStringStorage:
    printer_name = 'boost::wave::util::AllocatorStringStorage'
    min_supported_version = (1, 71, 0)
    max_supported_version = last_supported_boost_version
    template_name = 'boost::wave::util::AllocatorStringStorage'

    def __init__(self, val):
        self.val = val

    def display_hint(self):
        return 'string'

    def to_string(self):
        data = self.val['pData_'].dereference()

        str_begin = cast_array_to_pointer(data['buffer_'])
        str_end = data['pEnd_']
        str_length = str_end - str_begin
        return str_begin.string(length=str_length)


@add_printer
class BoostWaveFlexString:
    """Pretty Printer for Boost.Wave internal string"""
    printer_name = 'boost::wave::util::flex_string'
    min_supported_version = (1, 71, 0)
    max_supported_version = last_supported_boost_version
    template_name = 'boost::wave::util::flex_string'

    def __init__(self, val):
        self.val = val

    def display_hint(self):
        return 'string'

    def to_string(self):
        storage_type = self.val.type.template_argument(3)
        storage = self.val.cast(storage_type)
        return storage
```

Reply:
OK, I've revised using a modified version of your approach, and made a few other tweaks. Hope this passes muster.